python.d/ntp.chart.py

```diff
@@ -0,0 +1,381 @@
+
+                    match_peer_filter = self.regex_peer_filter.search(name)
+                    if match_peer_filter:
+                        continue
+
+                    if len(name) > len(domain) and name.endswith(domain):
+                        name = name[:-len(domain)]
+                except (IndexError, socket.error):
+                    self.error('Failed to reverse lookup address')
+
+            name = name.replace('.', '-')
+
+            peers.append(Peer(peer_id, name, request))
+
+        return peers
+
+    def check(self):
+        """
+        Checks if we can get valid systemvars.
+        If not, returns None to disable module.
+        """
+        self._parse_config()
+
+        self.request_systemvars = self.get_header(0, 'readvar')
+        self.request = self.request_systemvars
+        raw_systemvars = self._get_raw_data()
+
+        if not self.get_data_from_raw(raw_systemvars):
+            return None
+
+        self.create_charts()
+
+        return True
+
+    def _get_data(self):
+        """
+        Gets systemvars data on each update.
+        Gets peervars data for only one peer on each update.
+        Total amount of _get_raw_data invocations per update = 2
+        """
+        data = dict()
+
+        self.request = self.request_systemvars
+        raw_systemvars = self._get_raw_data()
+        data.update(self.get_data_from_raw(raw_systemvars))
+
+        if self.peers:
+            peer = next(self.peers)
+            self.request = peer.request
+            raw_peervars = self._get_raw_data()
+            data.update(self.get_data_from_raw(raw_peervars, peer))
+
+        if not data:
+            self.error("No data received")
+            return None
+
+        return data
+
+    def get_data_from_raw(self, raw, peer=None):
+        """
+        Extracts key=value pairs with float/integer from ntp response packet data.
+        """
+        data = dict()
+        try:
+            data_list = self.regex_data.findall(raw)
+
+            for data_point in data_list:
+                key, value = data_point
+                if peer:
+                    dimension = '_'.join([peer.name, key])
+                else:
+                    dimension = key
+                data[dimension] = int(float(value) * PRECISION)
+        except (ValueError, AttributeError, TypeError):
+            self.error("Invalid data received")
+            return None
+
+        # If peer returns no valid data, probably due to ntpd restart,
+        # then wait 10 seconds and re-initialize the peers and charts
+        if not data and peer:
+            self.error('Peer error: No data received')
+            self.peer_error += 1
+
+            if (self.peer_error * self.update_every) > 10:
+                self.error('Peer error count exceeded, re-creating charts.')
+                self.create_charts()
+                self.peer_error = 0
+                self.create()
```

Comment:
i `self.create()` and `self.create()` here. Not sure this is the right way to re-initialize the peers and charts.

Reply:
please explain why we need to re-initialize something.

Reply:
I am not sure as well. I just tried `self.create()` and it worked. What would be the right way to re-create or modify the charts?

I choose to re-creates all charts in this way to be as simple as possible. In theory only the peer_charts need to be modified/removed.

Reason:

This is needed for the `peer_charts` when `ntpd` is restarted. To get data for a peer, we need to make a request which includes the `peer_id` (or in ntp terms "association_id"). Unfortunately by design, these peer_ids (random integers) change on every `ntpd` restart, even if no configuration change was made. This means after a peer_error (when no data is returned) we need to request again all peer_ids. Then we need to re-create the peer charts, because if one or more pools are configured (or maybe if a configured server address is behind DNS round robin), there will most probably be new peers (with different ip addresses) after a restart, even if no configuration change was made.

Reply:
How is it possible to add new dimensions to charts at run-time?

After a restart, we could just request the new peer_ids and add the new dimensions (with possible new IP-addresses/host-names). Any obsolete dimensions should be cleaned up after some time, right?

Reply:
And yes, it should work after a restart of `ntpd.service`. The `system_vars` should always work and almost immediately. But the `peer_vars` need a few seconds to be initialized, because `ntpd` first has to poll its peers. If we request the new peer_ids before `ntpd` has properly initialized its peers, there will be no peer_charts shown anymore. This is why I choose to just wait 10s after a peer_error (but maybe there is a better way). We could also periodically get the peer_ids and check if some of the ids have changed.